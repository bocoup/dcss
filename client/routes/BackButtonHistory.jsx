import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';

class BackButtonHistory extends React.Component {
  constructor(props) {
    super(props);
    this.onPopState = this.onPopState.bind(this);
  }

  componentDidMount() {
    window.addEventListener('popstate', this.onPopState);
  }

  componentWillUnmount() {
    window.removeEventListener('popstate', this.onPopState);
  }

  onPopState() {
    window.location.href = this.props.location.pathname;
  }

  render() {
    return this.props.children;
  }
}

BackButtonHistory.propTypes = {
  children: PropTypes.array,
  location: PropTypes.object.isRequired
};

export default withRouter(BackButtonHistory);
