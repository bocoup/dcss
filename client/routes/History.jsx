import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';

class History extends Component {
  constructor(props) {
    super(props);

    this.onPopState = this.onPopState.bind(this);
  }

  static propTypes() {
    return {
      children: PropTypes.array,
      history: PropTypes.object
    };
  }

  componentDidMount() {
    window.addEventListener('popstate', this.onPopState);
  }

  componentWillUnmount() {
    window.removeEventListener('popstate', this.onPopState);
  }

  onPopState() {
    // console.log(event);
    // console.log(this.props.history);
    // console.log('onPopState');
    // this.props.history.goBack();
  }

  render() {
    return this.props.children;
  }
}

export default withRouter(History);
