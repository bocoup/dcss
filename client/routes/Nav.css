@media not all and (max-device-width: 767px) {
    .ui.fluid.button.nav__mobile {
        display: none;
    }
}

@media not all and (min-width: 767px) {
    .ui.fluid.button.nav__mobile {
        display: block;
    }
}

.ui.menu .item.navigation__menu-item-logo {
    padding: 0px;
}

.ui.menu .item.navigation__menu-item-logo a {
    padding: 10px;
    background-image: url('../static/images/MIT-TSL-Logo-Icons.png');
    background-position: center center;
    background-repeat: no-repeat;
    text-indent: -999em;
    min-width: 130px;
    background-position-x: center;
}
