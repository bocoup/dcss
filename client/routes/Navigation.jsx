import React, { useState } from 'react';
import { NavLink } from 'react-router-dom';
import { Button, Dropdown, Icon, Menu } from 'semantic-ui-react';

import ConfirmAuth from '@client/components/ConfirmAuth';
import ConfirmableLogoutMenuItem from '@client/components/Login/ConfirmableLogoutMenuItem';
import Session from '@client/util/session';

Session.timeout();

const MOBILE_WIDTH = 767;
const restrictedNav = [
    {
        text: 'Create a Moment',
        path: '/editor/new',
        permission: 'create_scenario'
    },
    {
        text: 'Cohorts',
        path: '/cohorts',
        permission: 'create_cohort'
    },
    {
        text: 'My Cohort',
        path: '/cohort',
        permission: 'view_invited_cohorts'
    },
    {
        text: 'Research',
        path: '/research',
        permission: 'view_run_data'
    },
    {
        text: 'Account Administration',
        path: '/account-administration',
        permission: 'edit_permissions'
    }
];

const Navigation = () => {
    const [menuExpanded, setMenuExpanded] = useState(
        !(window.innerWidth <= MOBILE_WIDTH)
    );
    window.onresize = () => {
        if (window.innerWidth > MOBILE_WIDTH && !menuExpanded) {
            // Make the menu appear if the user resizes the window to be wider
            setMenuExpanded(true);
        }
    };
    const authorizedNav = restrictedNav.map(
        ({ text, path, permission }, index) => {
            return (
                <ConfirmAuth key={index} requiredPermission={permission}>
                    <Menu.Item>
                        <NavLink to={path}>{text}</NavLink>
                    </Menu.Item>
                </ConfirmAuth>
            );
        }
    );
    const { username } = Session.getSession();

    return (
        <React.Fragment>
            <Button
                icon
                fluid
                labelPosition="left"
                basic
                className="nav__mobile"
                id="nav__mobile--button"
                onClick={() => setMenuExpanded(!menuExpanded)}
                aria-controls="nav"
            >
                <Icon name="content" />
                MENU
            </Button>
            {menuExpanded && (
                <Menu id="nav" className="nav" stackable>
                    <Menu.Item className="navigation__menu-item-logo">
                        <NavLink to="/">Teacher Moments</NavLink>
                    </Menu.Item>

                    <Menu.Menu>
                        <Dropdown
                            simple
                            item
                            text={<NavLink to="/">Moments</NavLink>}
                        >
                            <Dropdown.Menu>
                                {Session.isSessionActive() && (
                                    <React.Fragment>
                                        <ConfirmAuth requiredPermission="create_scenario">
                                            <Dropdown.Item>
                                                <NavLink
                                                    to={{
                                                        pathname: `/author/${username}`
                                                    }}
                                                >
                                                    My Moments
                                                </NavLink>
                                            </Dropdown.Item>
                                        </ConfirmAuth>
                                        <Dropdown.Item>
                                            <NavLink to="/continue">
                                                Continue Moments
                                            </NavLink>
                                        </Dropdown.Item>
                                    </React.Fragment>
                                )}
                                <Dropdown.Item>
                                    <NavLink to="/official">Official</NavLink>
                                </Dropdown.Item>
                                <Dropdown.Item>
                                    <NavLink to="/community">Community</NavLink>
                                </Dropdown.Item>
                            </Dropdown.Menu>
                        </Dropdown>
                    </Menu.Menu>

                    {authorizedNav}

                    {Session.isSessionActive() ? (
                        <ConfirmableLogoutMenuItem />
                    ) : (
                        <Menu.Item position="right">
                            <NavLink to="/login">Log in</NavLink>
                        </Menu.Item>
                    )}
                </Menu>
            )}
        </React.Fragment>
    );
};

export default Navigation;
