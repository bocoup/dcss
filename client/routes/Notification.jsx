import { SemanticToastContainer as Notification } from 'react-semantic-toasts';
import './Notification.css';

export default Notification;
