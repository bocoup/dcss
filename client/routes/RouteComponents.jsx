import React from 'react';
import { Route } from 'react-router-dom';
import PropTypes from 'prop-types';

import Editor from '@components/Editor';
import Login from '@components/Login';
import LoginRoutePromptModal from '@components/Login/LoginRoutePromptModal';
import ScenariosList from '@components/ScenariosList';

import Session from '@utils/session';

/**
 * Special case components for solving routing issues. Component state
 * doesn't update when navigating to exact same component.
 */

export const NewScenario = props => {
    return <Editor {...props} isNewScenario={true} />;
};

export const CopyScenario = props => {
    return <Editor {...props} isCopyScenario={true} />;
};

export const ScenariosListAll = props => {
    return <ScenariosList {...props} category="all" />;
};

export const ScenariosListAuthor = props => {
    return <ScenariosList {...props} category="author" />;
};

export const ScenariosListCommunity = props => {
    return <ScenariosList {...props} category="community" />;
};

export const ScenariosListOfficial = props => {
    return <ScenariosList {...props} category="official" />;
};

export const Logout = props => {
    return <Login {...props} mode="logout" />;
};

export const InterceptAnonymizableRoute = ({ children, ...rest }) => {
    return (
        <Route
            {...rest}
            render={props => {
                return Session.isSessionActive() ? (
                    children
                ) : (
                    <LoginRoutePromptModal {...props} />
                );
            }}
        />
    );
};

InterceptAnonymizableRoute.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]).isRequired
};

export const RedirectRouteForActiveSession = ({ children, ...rest }) => {
    return (
        <Route
            {...rest}
            render={() => {
                if (!Session.isSessionActive()) {
                    return children;
                }
                location.href = '/';
            }}
        />
    );
};

RedirectRouteForActiveSession.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]).isRequired
};

export const RedirectRouteForInactiveSession = ({ children, ...rest }) => {
    return (
        <Route
            {...rest}
            render={() => {
                if (Session.isSessionActive()) {
                    return children;
                }
                location.href = '/login';
            }}
        />
    );
};

RedirectRouteForInactiveSession.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]).isRequired
};
