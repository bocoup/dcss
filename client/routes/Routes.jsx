import React from 'react';
import { Route, Switch } from 'react-router-dom';
import PropTypes from 'prop-types';
import ScenariosList from '@client/components/ScenariosList';
import Editor from '@client/components/Editor';
import Facilitator from '@client/components/Facilitator';
import AccountAdmin from '@client/components/AccountAdmin';
import Login from '@client/components/Login';
import LoginRoutePromptModal from '@client/components/Login/LoginRoutePromptModal';
import CreateAccount from '@client/components/CreateAccount';
import CreateAnonymousAccount from '@client/components/CreateAccount/CreateAnonymousAccount';
import Run from '@client/components/Run';
import Researcher from '@client/components/Researcher';

import Session from '@client/util/session';

// Special case component for solving new scenario routing issue.
// Previously the route wouldn't update to the same Editor component.
const NewScenario = props => {
    return <Editor {...props} isNewScenario={true} />;
};

const Logout = props => {
    return <Login {...props} mode="logout" />;
};

const InterceptAnonymizableRoute = ({ children, ...rest }) => {
    return (
        <Route
            {...rest}
            render={props => {
                return Session.isSessionActive() ? (
                    children
                ) : (
                    <LoginRoutePromptModal {...props} />
                );
            }}
        />
    );
};

InterceptAnonymizableRoute.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]).isRequired
};

const Routes = () => {
    return (
        <Switch>
            <Route exact path="/" component={ScenariosList} />
            <Route exact path="/official" component={ScenariosList} />
            <Route exact path="/community" component={ScenariosList} />
            <InterceptAnonymizableRoute path="/run/:scenarioId/:activeSlideIndex">
                <Route component={Run} />
            </InterceptAnonymizableRoute>
            <InterceptAnonymizableRoute path="/run/:scenarioId">
                <Route component={Run} />
            </InterceptAnonymizableRoute>
            <Route exact path="/editor/new" component={NewScenario} />
            <Route path="/editor/:id" component={Editor} />
            <Route exact path="/cohorts" component={Facilitator} />
            <Route
                exact
                path="/account-administration"
                component={AccountAdmin}
            />
            <Route exact path="/logout" component={Logout} />
            <Route exact path="/login" component={Login} />
            <Route exact path="/login/new" component={CreateAccount} />
            <Route
                exact
                path="/login/anonymous"
                component={CreateAnonymousAccount}
            />
            <Route exact path="/researcher" component={Researcher} />
        </Switch>
    );
};

export default Routes;
