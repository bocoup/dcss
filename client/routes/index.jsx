import React, { Component } from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

import Navigation from './Navigation';
import Notification from './Notification';
import Routes from './Routes';
import './Nav.css';

import Session from '@client/util/session';

Session.timeout();

class App extends Component {
    constructor(props) {
        super(props);
    }
    render() {
        return (
            <Router>
                <Navigation />
                <Routes />
                <Notification />
            </Router>
        );
    }
}

export default App;
