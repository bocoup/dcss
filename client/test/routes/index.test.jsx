import {
  fetchImplementation,
  mounter,
  reduxer,
  snapshot,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';
import { cleanup, fireEvent, render } from '@testing-library/react';
import routes from '../../routes/index.jsx';

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const sharedProps = {
  history: {
    push() {}
  }
};

test('routes', () => {
  expect(routes).toBeDefined();
});

test('Snapshot 1', () => {
  const props = {
    ...sharedProps
  };
  const mounted = mounter(reduxer(routes, props, state))();
  expect(snapshot(mounted)).toMatchSnapshot();
});

/*{INJECTION}*/
