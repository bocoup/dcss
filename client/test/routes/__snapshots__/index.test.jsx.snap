// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1 1 1`] = `""`;

exports[`Snapshot 1 1 2`] = `""`;
